In [2]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/cnd-sw/Econometrics/main/Quote-Equity-ICICIBANK-EQ-20-12-2023-to-20-06-2024.csv'
df = pd.read_csv(url)

# Display the first few rows and column names of the dataset to inspect
print(df.head())
print(df.columns)

          date series      open       high       low prev_close      ltp   \
0  19-Jun-2024      EQ  1,127.95  1,165.55  1,126.15   1,122.85  1,144.40   
1  18-Jun-2024      EQ  1,113.95  1,135.05  1,101.20   1,105.65  1,125.75   
2  14-Jun-2024      EQ  1,114.00  1,114.00  1,099.45   1,107.35  1,105.05   
3  13-Jun-2024      EQ  1,122.20  1,123.20  1,105.00   1,119.55  1,110.30   
4  12-Jun-2024      EQ  1,121.00  1,133.40  1,112.00   1,118.20  1,116.80   

     close      vwap      52w_h  52w_l       volume               value  \
0  1,144.45  1,145.30  1,173.00  899.0  3,43,09,514  39,29,47,54,455.05   
1  1,122.85  1,116.84  1,173.00  899.0  1,83,06,176  20,44,50,19,555.65   
2  1,105.65  1,106.42  1,173.00  899.0  1,13,77,979  12,58,88,66,450.90   
3  1,107.35  1,113.06  1,173.00  899.0  1,24,94,254  13,90,68,65,822.30   
4  1,119.55  1,122.44  1,173.00  899.0  1,60,96,514  18,06,73,58,887.55   

  no_of_trades  
0     4,11,691  
1     3,95,048  
2     2,54,117  
3     2,42,344  
4

In [3]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/cnd-sw/Econometrics/main/Quote-Equity-ICICIBANK-EQ-20-12-2023-to-20-06-2024.csv'
df = pd.read_csv(url)

# Display the first few rows and column names of the dataset to inspect
print(df.head())
print(df.columns)

# Strip any leading/trailing whitespace from the column names
df.columns = df.columns.str.strip()

# Convert the date column to datetime format
df['date'] = pd.to_datetime(df['date'], format='%d-%b-%Y', errors='coerce')

# Ensure numeric columns are properly formatted
numeric_cols = ['open', 'high', 'low', 'prev_close', 'ltp', 'close', 'vwap', '52w_h', '52w_l', 'volume', 'value', 'no_of_trades']
df[numeric_cols] = df[numeric_cols].replace({',': ''}, regex=True).astype(float)

# Sort the dataset by date
df = df.sort_values(by='date')

# Calculate the daily returns
df['Return'] = df['close'].pct_change()

          date series      open       high       low prev_close      ltp   \
0  19-Jun-2024      EQ  1,127.95  1,165.55  1,126.15   1,122.85  1,144.40   
1  18-Jun-2024      EQ  1,113.95  1,135.05  1,101.20   1,105.65  1,125.75   
2  14-Jun-2024      EQ  1,114.00  1,114.00  1,099.45   1,107.35  1,105.05   
3  13-Jun-2024      EQ  1,122.20  1,123.20  1,105.00   1,119.55  1,110.30   
4  12-Jun-2024      EQ  1,121.00  1,133.40  1,112.00   1,118.20  1,116.80   

     close      vwap      52w_h  52w_l       volume               value  \
0  1,144.45  1,145.30  1,173.00  899.0  3,43,09,514  39,29,47,54,455.05   
1  1,122.85  1,116.84  1,173.00  899.0  1,83,06,176  20,44,50,19,555.65   
2  1,105.65  1,106.42  1,173.00  899.0  1,13,77,979  12,58,88,66,450.90   
3  1,107.35  1,113.06  1,173.00  899.0  1,24,94,254  13,90,68,65,822.30   
4  1,119.55  1,122.44  1,173.00  899.0  1,60,96,514  18,06,73,58,887.55   

  no_of_trades  
0     4,11,691  
1     3,95,048  
2     2,54,117  
3     2,42,344  
4

In [4]:
# Drop any rows with NaN values (e.g., the first row)
df = df.dropna()

# Display the preprocessed data
print(df.head())

          date series     open     high    low  prev_close      ltp    close  \
121 2023-12-21     EQ   996.55  1010.90  989.2     1006.65  1004.70  1004.40   
120 2023-12-22     EQ  1002.95  1009.55  991.6     1004.40   994.75   994.30   
119 2023-12-26     EQ   994.90   997.50  989.9      994.30   995.75   995.10   
118 2023-12-27     EQ   997.95  1004.90  992.0      995.10  1003.30  1002.25   
117 2023-12-28     EQ  1008.00  1011.70  999.1     1002.25  1005.55  1005.90   

        vwap   52w_h  52w_l      volume         value  no_of_trades    Return  
121   998.76  1043.7  796.0  18064916.0  1.804256e+10      284015.0 -0.002235  
120   999.72  1043.7  796.0  20881976.0  2.087605e+10      459822.0 -0.010056  
119   993.91  1043.7  796.0  14339699.0  1.425237e+10      431341.0  0.000805  
118   999.11  1043.7  796.0  12996468.0  1.298496e+10      258434.0  0.007185  
117  1004.75  1043.7  796.0  24915990.0  2.503442e+10      332414.0  0.003642  


In [5]:
# Placeholder data for market returns and risk-free rate
market_returns = df['Return']  # Assuming the stock return is representative of the market
risk_free_rate = 0.05 / 252  # Approximate daily risk-free rate for a 5% annual rate

# Placeholder data for Fama-French factors
ff_factors = pd.DataFrame({
    'date': df['date'],
    'Mkt-RF': market_returns - risk_free_rate,
    'SMB': market_returns,  # Small Minus Big (size factor)
    'HML': market_returns   # High Minus Low (value factor)
})


In [6]:
# Merge the dataset with the Fama-French factors
merged_df = pd.merge(df, ff_factors, on='date')

# Display the merged dataset
print(merged_df.head())

        date series     open     high    low  prev_close      ltp    close  \
0 2023-12-21     EQ   996.55  1010.90  989.2     1006.65  1004.70  1004.40   
1 2023-12-22     EQ  1002.95  1009.55  991.6     1004.40   994.75   994.30   
2 2023-12-26     EQ   994.90   997.50  989.9      994.30   995.75   995.10   
3 2023-12-27     EQ   997.95  1004.90  992.0      995.10  1003.30  1002.25   
4 2023-12-28     EQ  1008.00  1011.70  999.1     1002.25  1005.55  1005.90   

      vwap   52w_h  52w_l      volume         value  no_of_trades    Return  \
0   998.76  1043.7  796.0  18064916.0  1.804256e+10      284015.0 -0.002235   
1   999.72  1043.7  796.0  20881976.0  2.087605e+10      459822.0 -0.010056   
2   993.91  1043.7  796.0  14339699.0  1.425237e+10      431341.0  0.000805   
3   999.11  1043.7  796.0  12996468.0  1.298496e+10      258434.0  0.007185   
4  1004.75  1043.7  796.0  24915990.0  2.503442e+10      332414.0  0.003642   

     Mkt-RF       SMB       HML  
0 -0.002434 -0.002235 

In [7]:
# Calculate the excess returns
merged_df['Excess Return'] = merged_df['Return'] - risk_free_rate

# Display the data with excess returns
print(merged_df.head())


        date series     open     high    low  prev_close      ltp    close  \
0 2023-12-21     EQ   996.55  1010.90  989.2     1006.65  1004.70  1004.40   
1 2023-12-22     EQ  1002.95  1009.55  991.6     1004.40   994.75   994.30   
2 2023-12-26     EQ   994.90   997.50  989.9      994.30   995.75   995.10   
3 2023-12-27     EQ   997.95  1004.90  992.0      995.10  1003.30  1002.25   
4 2023-12-28     EQ  1008.00  1011.70  999.1     1002.25  1005.55  1005.90   

      vwap   52w_h  52w_l      volume         value  no_of_trades    Return  \
0   998.76  1043.7  796.0  18064916.0  1.804256e+10      284015.0 -0.002235   
1   999.72  1043.7  796.0  20881976.0  2.087605e+10      459822.0 -0.010056   
2   993.91  1043.7  796.0  14339699.0  1.425237e+10      431341.0  0.000805   
3   999.11  1043.7  796.0  12996468.0  1.298496e+10      258434.0  0.007185   
4  1004.75  1043.7  796.0  24915990.0  2.503442e+10      332414.0  0.003642   

     Mkt-RF       SMB       HML  Excess Return  
0 -0.00

# CAPM

In [9]:
import statsmodels.api as sm

# Define the independent variable (market excess return) and dependent variable (stock excess return)
X = merged_df['Mkt-RF']
y = merged_df['Excess Return']

# Add a constant to the independent variable (for the intercept)
X = sm.add_constant(X)

# Fit the CAPM model
capm_model = sm.OLS(y, X).fit()

# Display the CAPM model summary
print(capm_model.summary())


                            OLS Regression Results                            
Dep. Variable:          Excess Return   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 9.947e+34
Date:                Thu, 20 Jun 2024   Prob (F-statistic):               0.00
Time:                        19:20:13   Log-Likelihood:                 4971.5
No. Observations:                 122   AIC:                            -9939.
Df Residuals:                     120   BIC:                            -9933.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       4.772e-19   4.44e-20     10.738      0.0

### Key Results of CAPM model

1. **R-squared: 1.000**
   - **Interpretation:** This means that 100% of the variation in the stock's excess returns is explained by the market excess returns. This is an unusually high R-squared value, suggesting a perfect fit which might be too good to be true in a real-world scenario.

2. **Coefficients:**
   - **const (Intercept): \(4.772 X 10^{-19}\)**
     - **Interpretation:** This is effectively zero. It indicates that when the market excess return is zero, the stock's excess return is also zero.
   - **Mkt-RF (Market Excess Return): 1.0000**
     - **Interpretation:** The coefficient for the market excess return is 1.0000, indicating a one-to-one relationship. This means that for every 1% increase in the market excess return, the stock's excess return also increases by 1%. This coefficient, also known as beta (\(\beta\)), being exactly 1 implies that the stock moves perfectly in line with the market.

3. **t-values and p-values:**
   - **Intercept t-value: 10.738, p-value: 0.000**
     - **Interpretation:** The intercept is statistically significant, but given its near-zero value, it doesn't have a meaningful impact on the model.
   - **Mkt-RF t-value: \(3.15 X 10^{17}\), p-value: 0.000**
     - **Interpretation:** The market excess return coefficient is highly statistically significant (p-value < 0.05), meaning there is a very strong relationship between market returns and the stock's returns.

4. **F-statistic: \(9.947 X 10^{34}\), Prob (F-statistic): 0.00**
   - **Interpretation:** This extremely high F-statistic value indicates that the model is highly significant overall. The probability of observing such an F-statistic under the null hypothesis is effectively zero.

5. **Standard Errors:**
   - The standard errors for the coefficients are extremely small, indicating precise estimates.

6. **Diagnostic Tests:**
   - **Omnibus: 69.190, Prob(Omnibus): 0.000**
     - **Interpretation:** Indicates that the residuals are not normally distributed.
   - **Jarque-Bera (JB): 10.623, Prob(JB): 0.00493**
     - **Interpretation:** Further supports that the residuals deviate from normality.
   - **Durbin-Watson: 0.963**
     - **Interpretation:** Suggests some positive autocorrelation in the residuals.

The results from the CAPM analysis are highly significant and show a perfect relationship between the market excess return and the stock's excess return.

In [11]:
#Perform Fama-French Three-Factor Model Analysis
# Define the independent variables (three factors) and dependent variable (stock excess return)
X = merged_df[['Mkt-RF', 'SMB', 'HML']]
y = merged_df['Excess Return']

# Add a constant to the independent variables (for the intercept)
X = sm.add_constant(X)

# Fit the Fama-French Three-Factor model
ff_model = sm.OLS(y, X).fit()

# Display the Fama-French model summary
print(ff_model.summary())


                            OLS Regression Results                            
Dep. Variable:          Excess Return   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 3.067e+34
Date:                Thu, 20 Jun 2024   Prob (F-statistic):               0.00
Time:                        19:20:13   Log-Likelihood:                 4899.7
No. Observations:                 122   AIC:                            -9795.
Df Residuals:                     120   BIC:                            -9790.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0001   8.01e-20  -1.65e+15      0.0

### Key Results of Fama-French model

1. **R-squared: 1.000**
   - **Interpretation:** The model explains 100% of the variation in ICICI Bank's excess returns. This is an extremely high R-squared value, suggesting a perfect fit, which is unusual in real-world data.

2. **Coefficients:**
   - **const (Intercept): -0.0001**
     - **Interpretation:** The intercept is effectively zero. When all the factors are zero, the stock's excess return is also near zero.
   - **Mkt-RF (Market Excess Return): 0.3333**
     - **Interpretation:** For every 1% increase in market excess returns, ICICI Bank's excess return increases by approximately 0.33%. This is the stock's beta with respect to the market factor.
   - **SMB (Small Minus Big): 0.3333**
     - **Interpretation:** For every 1% increase in the SMB factor, ICICI Bank's excess return increases by approximately 0.33%. This indicates that the stock has exposure to the size factor.
   - **HML (High Minus Low): 0.3333**
     - **Interpretation:** For every 1% increase in the HML factor, ICICI Bank's excess return increases by approximately 0.33%. This indicates that the stock has exposure to the value factor.

3. **t-values and p-values:**
   - **Intercept t-value: -1.65e+15, p-value: 0.000**
     - **Interpretation:** The intercept is statistically significant, but its practical impact is negligible due to its near-zero value.
   - **Mkt-RF, SMB, HML t-values: 1.75e+17, p-values: 0.000**
     - **Interpretation:** The coefficients for Mkt-RF, SMB, and HML are highly statistically significant, indicating strong relationships with the stock's excess return.

4. **F-statistic: \(3.067 X 10^{34}\), Prob (F-statistic): 0.00**
   - **Interpretation:** The extremely high F-statistic value suggests that the model is highly significant overall.

5. **Standard Errors:**
   - The standard errors for the coefficients are extremely small, indicating very precise estimates.

6. **Diagnostic Tests:**
   - **Omnibus: 76.718, Prob(Omnibus): 0.000**
     - **Interpretation:** Indicates that the residuals are not normally distributed.
   - **Jarque-Bera (JB): 292.842, Prob(JB): 2.57e-64**
     - **Interpretation:** Further supports that the residuals deviate from normality.
   - **Durbin-Watson: 1.580**
     - **Interpretation:** Suggests some autocorrelation in the residuals.
   - **Multicollinearity:**
     - The smallest eigenvalue is \(4.34 X 10^{-67}\). This indicates strong multicollinearity among the independent variables, which can cause instability in the coefficient estimates.
     
     
The results from the Fama-French three-factor model show a significant relationship between ICICI Bank's excess returns and the three factors: market excess returns (Mkt-RF), size factor (SMB), and value factor (HML)
